packages/flutter_tools/lib/src/crash_reporting.dart

```diff
@@ -133,4 +151,183 @@ class CrashReportSender {
+      req.fields['error_runtime_type'] = _kEngineTypeId;
+
+      if (engineCrash.lineContaining('backtrace') != -1) {
+        req.files.add(http.MultipartFile.fromString(
+          _kEngineStackTraceFileField,
+          engineCrash.squashBacktrace(
+            startLine: engineCrash.lineContaining('backtrace') + 1,
+          ),
+          filename: _kEngineStackTraceFileField,
+        ));
+
+        req.files.add(http.MultipartFile.fromString(
+          _kEngineRegistersFileField,
+          engineCrash.squashBacktrace(
+            startLine: engineCrash.lineContaining('backtrace') - 5,
+            endLine: engineCrash.lineContaining('backtrace')
+          ),
+          filename: _kEngineRegistersFileField,
+        ));
+
+        req.files.add(http.MultipartFile.fromString(
+          _kEngineErrorFileField,
+          engineCrash.squashBacktrace(
+            endLine: engineCrash.lineContaining('backtrace: ') - 4,
+          ),
+          filename: _kEngineErrorFileField,
+        ));
+      } else {
+        // Malformed error with no backtrace. Report entire error.
+        req.files.add(http.MultipartFile.fromString(
+          _kEngineErrorFileField,
+          engineCrash.squashBacktrace(),
+          filename: _kEngineErrorFileField,
+        ));
+      }
+
+      final http.StreamedResponse resp = await _client.send(req);
+
+      if (resp.statusCode == 200) {
+        final String reportId = await http.ByteStream(resp.stream)
+            .bytesToString();
+        printStatus('Crash report sent (report ID: $reportId)');
+      } else {
+        printError('Failed to send crash report. Server responded with HTTP status code ${resp.statusCode}');
+      }
+    } catch (sendError, sendStackTrace) {
+      if (sendError is SocketException) {
+        printError('Failed to send crash report due to a network error: $sendError');
+      } else {
+        // If the sender itself crashes, just print. We did our best.
+        printError('Crash report sender itself crashed: $sendError\n$sendStackTrace');
+      }
+    }
+    completer.complete();
+  }
+}
+
+// Holds the parsed data (from logs) of an engine crash.
+class EngineCrash {
+
+  EngineCrash() {
+    error = '';
+    _backtrace = <String>[];
+  }
+
+  String error;
+
+  // Lines that make up the backtrace.
+  List<String> _backtrace;
+
+  // Completes when the report is fully submitted or skipped. Is null when
+  // no report attempt has been made.
+  Future<void> reportStatus;
+
+  void addTraceLine(String line) {
+    _backtrace.add(line);
+  }
+
+  String squashBacktrace({int startLine = 0, int endLine = -1}) {
+    if (endLine < 0) {
+      endLine = _backtrace.length;
+    }
+    String out = '';
+    for (; startLine < endLine; startLine++) {
+      out += _backtrace[startLine] + '\n';
+    }
+    return out;
+  }
+
+  int lineContaining(String str) {
+    for (int i = 0; i < _backtrace.length; i++) {
+      final String line = _backtrace[i];
```

Comment:
Nit: `line` seems to be unnecessary. Maybe `if (_backtrace[i].contains(str)) {` is cleaner?